### ISSUE TITLE
query with GSI return nothing (when using ALL_PROJECTED_ATTRIBUTES)

### ISSUE THREAD

**Issue by swimablefish** [mhart/dynalite#72]
Here is the index info in the table:

```
 {u'IndexArn': u'arn:aws:dynamodb:us-east-1:000000000000:table/table_1/index/index1',
                       u'IndexName': u'index1',
                       u'IndexSizeBytes': 121919705612,
                       u'IndexStatus': u'ACTIVE',
                       u'ItemCount': 1126912705,
                       u'KeySchema': [{u'AttributeName': u'tx_h',
                                                    u'KeyType': u'HASH'},
                                                  {u'AttributeName': u's',
                                                    u'KeyType': u'RANGE'}],
                        u'Projection': {u'NonKeyAttributes': [u'n'],
                                                u'ProjectionType': u'INCLUDE'},
```

And I use boto3 to query the table, it return nothing. If I query the real dynamodb, it can return the relevant items:

```python
resp = client.query(TableName='table_1',
                    IndexName='index1',
                    Select='ALL_PROJECTED_ATTRIBUTES',
                    KeyConditionExpression='tx_h = :tx_h',
                    ExpressionAttributeValues={":tx_h": {"S": 'xxxxxxx'}})
```

**Comment by mhart** [comment]
Hmmm, it seems that `ALL_PROJECTED_ATTRIBUTES` isn't implemented yet – will change the title of this and update you when I've gotten it working

**Comment by swimablefish** [comment]
even don't use `Select` parameter still return nothing

**Comment by swimablefish** [comment]
Sorry, It seems our program's problem, which is used to create table. Although it create GSI, it doesn't take effect. When I use boto3 to create tables, it works. The query can return the data I expect.

**Comment by swimablefish** [comment]
Clarify the scenario. It happens when there are multiple GSIs in a table. The query only returns result when use the first GSI,  or it will return nothing

**Comment by mhart** [comment]
@swimablefish can you show me the full table definition you're using?

**Comment by swimablefish** [comment]
Here is the code. If I want to query index3, I need to put index3 as the first gsi

```python
client.create_table(
    TableName = 'test_table',
    AttributeDefinitions=[
        {'AttributeName': 'i', 'AttributeType': 'B'},
        {'AttributeName': 'tx_j', 'AttributeType': 'S'},
        {'AttributeName': 'a_h', 'AttributeType': 'S'},
        {'AttributeName': 'tx_h', 'AttributeType': 'S'},
        {'AttributeName': 's', 'AttributeType': 'N'},
        {'AttributeName': 'r', 'AttributeType': 'N'},],
    KeySchema=[
        {'AttributeName': 'i', 'KeyType': 'HASH'},
        {'AttributeName': 'tx_j', 'KeyType': 'RANGE'},],
    GlobalSecondaryIndexes=[
        {
            'IndexName': 'index1',
            'KeySchema': [
                {'AttributeName': 'a_h', 'KeyType': 'HASH'},
                {'AttributeName': 'i', 'KeyType': 'RANGE'}],
            'Projection': {
                'ProjectionType': 'INCLUDE',
                'NonKeyAttributes': ['fs', 'n']
            },
            'ProvisionedThroughput': {'ReadCapacityUnits': 123, 'WriteCapacityUnits': 123}
        },
        {
            'IndexName': 'index2',
            'KeySchema': [
                {'AttributeName': 'a_h', 'KeyType': 'HASH'},
                {'AttributeName': 'r', 'KeyType': 'RANGE'}],
            'Projection': {
                'ProjectionType': 'INCLUDE',
                'NonKeyAttributes': ['fs']
            },
            'ProvisionedThroughput': {'ReadCapacityUnits': 123, 'WriteCapacityUnits': 123}
        },
        {
            'IndexName': 'index3',
            'KeySchema': [
                {'AttributeName': 'tx_h', 'KeyType': 'HASH'},
                {'AttributeName': 's', 'KeyType': 'RANGE'}],
            'Projection': {
                'ProjectionType': 'INCLUDE',
                'NonKeyAttributes': ['n']
            },
            'ProvisionedThroughput': {'ReadCapacityUnits': 123, 'WriteCapacityUnits': 123}
        },
    ],
    ProvisionedThroughput={'ReadCapacityUnits': 123, 'WriteCapacityUnits': 123}
)
```

**Comment by swimablefish** [comment]
found the problem:
1.  in `getIndexActions`, local var `item` is modified which is incomplete to the other GSIs besides the first GSI
2. some data don't have the range key when insert

**Comment by mhart** [comment]
@swimablefish sorry, I don't understand what you mean with 1 – can you explain?

**Comment by swimablefish** [comment]
https://github.com/mhart/dynalite/blob/master/db/index.js#L908,L922

If there are two GSIs: `hash=a, range=b`, `hash=c, range=b`, initial item = `{a: 1, b: 2, c: 3}`.
After the first round, the item will be `{a:1, b:2}`, then the second round it can't find key `c`, so it won't create the index data for the second GSI

**Comment by mhart** [comment]
@swimablefish ahhhh, gotcha – right, yeah, stupid reuse of the variable there. Nice find!

**Comment by mhart** [comment]
@swimablefish I just released v1.1.2 which address the variable reuse problem.

Can you explain the other issue you're seeing? I noticed that you opened #74 but unfortunately it's failing a bunch of tests. If you can add a test case in there that illustrates the problem you're seeing, that would be helpful

**Comment by swimablefish** [comment]
@mhart our data might not have the range key when is inserted, so they will be filtered by current logic, so I only check the hash key
https://github.com/mhart/dynalite/blob/master/db/index.js#L911  ,
https://github.com/mhart/dynalite/blob/master/db/index.js#L924 ,

and 
https://github.com/mhart/dynalite/blob/master/db/index.js#L243

In our cases, the `item[attr]` might be `undefined`

okay, I will check the test case later.
